lab-ui-manager/operator/operator.py

```diff
@@ -25,8 +28,29 @@
 custom_objects_api = kubernetes.client.CustomObjectsApi()
 j2env = jinja2.Environment(trim_blocks = True)
 
+class InfiniteRelativeBackoff:
+    def __init__(self, n=2, maximum=60):
+        self.n = n
+        self.maximum = maximum
+
+    def __iter__(self):
+        prev_t = []
+        max_age = self.maximum * math.ceil(math.log(self.maximum) / math.log(self.n))
+        while True:
+            t = time.monotonic()
+            prev_t = [p for p in prev_t if t - p < max_age]
+            delay = self.n ** len(prev_t)
+            prev_t.append(t)
```

Comment:
Why do you need to compute prev_t? that's a loop computation for each iteration. If it's infinite retry, how big can it grow?
Why append prev t to a list? can you just keep the last? Why do you to capture time here?

```python

    def __iter__(self):
            c = 0
            while True:
                delay = self.n ** c
                yield delay if delay < self.maximum else self.maximum
                c = c + 1
```

Reply:
Hrm... I was working with the assumption that the iterator is not reset by a single success, but now thinking it through you may be right that we can simplify the logic.

Reply:
hrm...
https://github.com/nolar/kopf/blob/main/kopf/_cogs/clients/api.py#L67-L70
```
    backoffs = settings.networking.error_backoffs
    backoffs = backoffs if isinstance(backoffs, collections.abc.Iterable) else [backoffs]
    count = len(backoffs) + 1 if isinstance(backoffs, collections.abc.Sized) else None
    backoff: Optional[float]
```

So, the kopf uses the same instance of whatever is set for `error_backoffs`. So using the simpler logic, it would never reset no matter how many successes were encountered. I suspected this might be the case. That's why the logic I wrote just maintains its own internal state. It checks how many times it has iterated error retries within the maximum backoff period and scales the delay based on that.

Reply:
Okay, much simpler now I think. Please have another look.